### DISCUSSION TITLE
Undefined References

### DISCUSSION THREAD

**Discussion by ThothK** [fltk/fltk#1067] [Q&A] [answered]
Hello,

I am experiencing a series of linker errors. There are a number of undefined references.
I am using FLTK 1.3. 
Please take a look and let me know any useful suggestions.

usr/bin/ld: /Libraries/FLTK013/libfltk.a(Fl_Double_Window.o): in function `Fl_Double_Window::flush(int)':
./src/Fl_Double_Window.cxx:443: undefined reference to `XdbeSwapBuffers'
/usr/bin/ld: ./src/Fl_Double_Window.cxx:406: undefined reference to `XdbeAllocateBackBufferName'
/usr/bin/ld: ./src/Fl_Double_Window.cxx:45: undefined reference to `XdbeQueryExtension'

Some of the other undefined reference are:
`XdbeGetVisualInfo', `XdbeFreeVisualInfo', `XdbeDeallocateBackBufferName', `XdbeDeallocateBackBufferName'
XFixesQueryExtension', `XcursorImageCreate', `XcursorImageLoadCursor', `XcursorImageDestroy', `XFixesSelectSelectionInput', `XftTextExtents32', `XftFontOpenXlfd', `XftFontMatch', `XftFontOpenPattern', `XftFontOpen', `XftTextExtents32', `XftDrawChange', `XftDrawSetClip', `XftDrawString32', `XftDrawCreate',  `XftTextExtents32', `XftDrawChange', `XftDrawSetClip', `XftDrawString32' ,`XftDrawCreate', `XftDrawChange'
`XineramaIsActive', `XineramaQueryScreens'

I am not seeing enough of a pattern to guess at what I should install. 

Thanks

**Comment by MatthiasWM** [comment]
- Did you build FLTK from scratch or did you install it with apt-get or similar?
- Are you building FLTK demos from teh FLTK archive, or are you building another app?
- What did you use to build that app? How did you determine the libraries to link with? `fltk-config`? CMake?

**Comment by ThothK** [comment]
Hello Matthias,

Thank you for answering.

I did not build FLTK from scratch. I get failures during the configure and make process.
I installed FLTK on a Debian 10 system using apt. I also installed FLTK on a Debian 12 system using apt. The newer system is where I am having errors.

apt install libfltk1.3-dev fluid | tee -a /Logs/Programs/fltk_'date +%Y%m%d'.txt
apt install libx11-dev libxrender-dev | tee -a /Logs/Programs/x_dev_'date +%Y%m%d'.txt

FLTK-Config reports
SurfaceGo3: ~: fltk-config --api-version
1.3

I do not recall what fltk-config reported for dependencies. My history shows I did this in 2022 and I did not make notes about what it generated.

I am using Code::Blocks to build the program on both computers. Code::Blocks is using GNU tools to compile and link with.

**Comment by chezbel** [comment]
I'm using FLTK 1.4 so I'm not sure if the set of dependencies is the same, but README.Unix.txt lists the following as required;
**libxft-dev** seems like it might be missing since a lot of your undefines are "Xft...". Maybe some others are missing as well.
```
g++
gdb
git
make
cmake      
autoconf   
libx11-dev
libglu1-mesa-dev
libxft-dev
libxcursor-dev
```
These ones are listed as recommended
```
libasound2-dev
freeglut3-dev
libcairo2-dev
libfontconfig1-dev
libglew-dev
libjpeg-dev
libpng-dev
libpango1.0-dev
libxinerama-dev
```

**Comment by ThothK** [comment, marked as answer]
Hi Chezbel,

Thank you for answering.

I reinstalled libx11-dev and installed libglu1-mesa-dev, libxft-dev, and libxcursor-dev.

For now, the unresolved references are gone. Hopefully they don't resurface after I resolve the current udev references.
